frontend/utils/distribution/group-income-distribution.js

```diff
@@ -1,40 +1,253 @@
-  // and historical groupProfiles. Since together these change across multiple
-  // locations in the code, it involves less 'code smell' to do it this way.
-  // see historical/group.js for the ugly way of doing it.
-  const mincomeAmount = getters.groupMincomeAmount
-  const groupProfiles = getters.groupProfiles
-  const currentIncomeDistribution = []
-  for (const username in groupProfiles) {
-    const profile = groupProfiles[username]
-    const incomeDetailsType = profile && profile.incomeDetailsType
-    if (incomeDetailsType) {
-      const adjustment = incomeDetailsType === 'incomeAmount' ? 0 : mincomeAmount
-      const amount = adjustment + profile[incomeDetailsType]
-      currentIncomeDistribution.push({
-        name: username,
-        amount: saferFloat(amount)
-      })
+export function dataToEvents (monthstamp, data) {
+  const mapUser = ({ withDate }) => ([name, profile]) => ({
+    name,
+    ...(withDate && { date: profile.joinedDate }),
+    ...(profile.incomeDetailsType === 'pledgeAmount'
+      ? { have: profile.pledgeAmount }
+      : { need: data.mincomeAmount - profile.incomeAmount })
+  })
+
+  const paymentEvents = []
+  if (data.adjustWith) {
+    const thisMonth = data.adjustWith.monthlyPayments[monthstamp]
+    if (thisMonth) {
+      for (const [from, toPayments] of Object.entries(thisMonth.paymentsFrom)) {
+        for (const [to, payments] of Object.entries(toPayments)) {
+          for (const paymentName of payments) {
+            const payment = data.adjustWith.payments[paymentName]
+            const amount = payment.amount * payment.exchangeRate * thisMonth.mincomeExchangeRate
+            const date = payment.createdDate
+            paymentEvents.push({ type: 'payment', from, to, amount, date })
+          }
+        }
+      }
+    }
+  }
+
+  const joinEvents = (Object.entries(data.groupProfiles)
+    .filter(([,profile]) => profile.joinedDate.startsWith(monthstamp))
+    .map(mapUser({ withDate: true }))
+    .map(user => { user.type = 'join'; return user }))
+
+  const events = (paymentEvents
+    .concat(joinEvents)
+    .sort((a, b) => a.date < b.date ? -1 : 1)
+    .map(event => { delete event.date; return event }))
+
+  const members = (Object.entries(data.groupProfiles)
+    .filter(([,profile]) => profile.joinedDate < monthstamp)
+    .map(mapUser({ withDate: false })))
+
+  return {
+    haves: members.filter(user => user.have !== undefined && user.have > 0),
+    needs: members.filter(user => user.need !== undefined),
+    events
+  }
+}
+
+/*
+JS is giving us 20 * (10 / 18) * (18 / 50) = 3.9999999999999996 instead of 4
+*/
+
+function distibuteFromHavesToNeeds({ haves, needs }) {
+  const totalHave = haves.reduce((a, b) => a + b.have, 0)
+  const totalNeed = needs.reduce((a, b) => a + b.need, 0)
+  const totalPercent = Math.min(1, totalHave / totalNeed)
+
+  for (const have of haves) have.percent = have.have / totalHave
```

Comment:
Does this need its own loop or can it be done inside of the other `for (const have of haves)` loop down below on line 66?

Reply:
We could recalculate it every time, sure. But I did this partly to help with debug-logging so that I could see the variables in the console.